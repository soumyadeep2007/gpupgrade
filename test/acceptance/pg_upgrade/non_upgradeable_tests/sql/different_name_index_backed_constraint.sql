-- Copyright (c) 2017-2021 VMware, Inc. or its affiliates
-- SPDX-License-Identifier: Apache-2.0

-- For unique or primary key constraints, the index name is auto generated, and
-- if the default index name is already taken by other objects, an incremental
-- number is appended to the index name. This means that pg_upgrade cannot
-- upgrade a cluster containing indexes of such type, they must be handled
-- manually before/after the upgrade. Although, the issue exists only with such
-- indexes, we wholesale ban upgrading of unique or primary key constraints.
-- Such, constraints must be dropped before the upgrade, and can be recreated
-- after the upgrade.

--------------------------------------------------------------------------------
-- Create and setup non-upgradeable objects
--------------------------------------------------------------------------------
CREATE TYPE table_with_unique_constraint_author_key AS (dummy int);
CREATE TYPE table_with_unique_constraint_author_key1 AS (dummy int);
-- If constraint is named then it leads to table which cannot be recreated from
-- pg_dump due to mismatch between backed index name and constraint name.
CREATE TABLE table_with_unique_constraint (author int, title int, CONSTRAINT table_with_unique_constraint_uniq_au_ti UNIQUE (author, title)) DISTRIBUTED BY (author);
DROP TYPE table_with_unique_constraint_author_key, table_with_unique_constraint_author_key1;
ALTER TABLE table_with_unique_constraint ADD PRIMARY KEY (author, title);
INSERT INTO table_with_unique_constraint VALUES(1, 1);
INSERT INTO table_with_unique_constraint VALUES(2, 2);

CREATE TYPE table_with_primary_constraint_pkey AS (dummy int);
CREATE TYPE table_with_primary_constraint_pkey1 AS (dummy int);
CREATE TABLE table_with_primary_constraint (author int, title int, CONSTRAINT table_with_primary_constraint_au_ti PRIMARY KEY (author, title)) DISTRIBUTED BY (author);
DROP TYPE table_with_primary_constraint_pkey, table_with_primary_constraint_pkey1;
ALTER TABLE table_with_primary_constraint ADD UNIQUE (author, title);
INSERT INTO table_with_primary_constraint VALUES(1, 1);
INSERT INTO table_with_primary_constraint VALUES(2, 2);

--------------------------------------------------------------------------------
-- Assert that pg_upgrade --check correctly detects the non-upgradeable objects
--------------------------------------------------------------------------------
!\retcode gpupgrade initialize --source-gphome="${GPHOME_SOURCE}" --target-gphome=${GPHOME_TARGET} --source-master-port=${PGPORT} --disk-free-ratio 0 --automatic;
! cat ${GPUPGRADE_HOME}/pg_upgrade/seg-1/unique_primary_key_constraint.txt | sort;

--------------------------------------------------------------------------------
-- Workaround to unblock upgrade
--------------------------------------------------------------------------------
ALTER TABLE table_with_unique_constraint DROP CONSTRAINT table_with_unique_constraint_uniq_au_ti;
ALTER TABLE table_with_primary_constraint DROP CONSTRAINT table_with_primary_constraint_au_ti;
