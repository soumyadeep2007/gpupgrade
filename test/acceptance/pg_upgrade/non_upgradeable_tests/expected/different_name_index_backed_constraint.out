-- Copyright (c) 2017-2021 VMware, Inc. or its affiliates
-- SPDX-License-Identifier: Apache-2.0

-- For unique or primary key constraints, the index name is auto generated, and
-- if the default index name is already taken by other objects, an incremental
-- number is appended to the index name. This means that pg_upgrade cannot
-- upgrade a cluster containing indexes of such type, they must be handled
-- manually before/after the upgrade. Although, the issue exists only with such
-- indexes, we wholesale ban upgrading of unique or primary key constraints.
-- Such, constraints must be dropped before the upgrade, and can be recreated
-- after the upgrade.

--------------------------------------------------------------------------------
-- Create and setup non-upgradeable objects
--------------------------------------------------------------------------------
CREATE TYPE table_with_unique_constraint_author_key AS (dummy int);
CREATE
CREATE TYPE table_with_unique_constraint_author_key1 AS (dummy int);
CREATE
-- If constraint is named then it leads to table which cannot be recreated from
-- pg_dump due to mismatch between backed index name and constraint name.
CREATE TABLE table_with_unique_constraint (author int, title int, CONSTRAINT table_with_unique_constraint_uniq_au_ti UNIQUE (author, title)) DISTRIBUTED BY (author);
CREATE
DROP TYPE table_with_unique_constraint_author_key, table_with_unique_constraint_author_key1;
DROP
ALTER TABLE table_with_unique_constraint ADD PRIMARY KEY (author, title);
ALTER
INSERT INTO table_with_unique_constraint VALUES(1, 1);
INSERT 1
INSERT INTO table_with_unique_constraint VALUES(2, 2);
INSERT 1

CREATE TYPE table_with_primary_constraint_pkey AS (dummy int);
CREATE
CREATE TYPE table_with_primary_constraint_pkey1 AS (dummy int);
CREATE
CREATE TABLE table_with_primary_constraint (author int, title int, CONSTRAINT table_with_primary_constraint_au_ti PRIMARY KEY (author, title)) DISTRIBUTED BY (author);
CREATE
DROP TYPE table_with_primary_constraint_pkey, table_with_primary_constraint_pkey1;
DROP
ALTER TABLE table_with_primary_constraint ADD UNIQUE (author, title);
ALTER
INSERT INTO table_with_primary_constraint VALUES(1, 1);
INSERT 1
INSERT INTO table_with_primary_constraint VALUES(2, 2);
INSERT 1

--------------------------------------------------------------------------------
-- Assert that pg_upgrade --check correctly detects the non-upgradeable objects
--------------------------------------------------------------------------------
!\retcode gpupgrade initialize --source-gphome="${GPHOME_SOURCE}" --target-gphome=${GPHOME_TARGET} --source-master-port=${PGPORT} --disk-free-ratio 0 --automatic;
-- start_ignore
-- end_ignore
(exited with code 1)
! cat ${GPUPGRADE_HOME}/pg_upgrade/seg-1/unique_primary_key_constraint.txt | sort;
Constraint name "table_with_primary_constraint_au_ti" does not match index name "table_with_primary_constraint_pkey2"
Constraint name "table_with_unique_constraint_uniq_au_ti" does not match index name "table_with_unique_constraint_author_key2"
Database:  isolation2test


--------------------------------------------------------------------------------
-- Workaround to unblock upgrade
--------------------------------------------------------------------------------
ALTER TABLE table_with_unique_constraint DROP CONSTRAINT table_with_unique_constraint_uniq_au_ti;
ALTER
ALTER TABLE table_with_primary_constraint DROP CONSTRAINT table_with_primary_constraint_au_ti;
ALTER
